compiler/rustc_ast/src/ast.rs

```diff
@@ -2810,16 +2810,19 @@ pub struct FnHeader {
     pub constness: Const,
     /// The `extern` keyword and corresponding ABI string, if any
     pub ext: Extern,
+    /// The `gen` keyword, if any
+    pub genness: Gen,
```

Comment:
I think asyncness and genness should be unified into one. There's no need to allocate two node-ids for `async gen` fns. It should just be:

```
enum ??Name?? {
  Async(),
  Gen(..),
  AsyncGen(..),
  None,
}
```

Reply:
Maybe just uplift `FnReturnTransform` into this crate and use that directly, since that seems to capture the right semantics automatically.

Reply:
I like unifying `Async` and `Gen` because that will simplify a lot of the later code (for example, all the places where I have `match (asyncness, genness) { (Async::Yes { .. }, _) | (_, Gen::Yes { .. }) => ... }` and it makes `async gen` impossible for now.

I did it this way originally because it seemed like the for AST we wanted a field per qualifier, but I'm happy to unify them.

Reply:
Done!